test/inference/gibbs_conditional.jl

```diff
@@ -0,0 +1,113 @@
+            m_n = N * x_bar / (N + 1)
+            λ_n = c.λ * (N + 1)
+            σ_n = sqrt(1 / λ_n)
+            return Normal(m_n, σ_n)
+        end
+
+        function cond_λ(c)
+            N, x_bar, s2 = gdemo_statistics(x)
+            α_n = α_0 + (N - 1) / 2 + 1
+            β_n = s2 * N / 2 + c.m^2 / 2 + inv(θ_0)
+            return Gamma(α_n, inv(β_n))
+        end
+
+        Random.seed!(100)
+
+        alg = Gibbs(
+            GibbsConditional(:m, cond_m),
+            GibbsConditional(:λ, c -> Normal(λ_true, 0)))
+        chain = sample(inverse_gdemo(x), alg, 10_000)
+        check_numerical(chain, [:m, :λ], [m_true, λ_true], atol=0.2)
+
+        alg = Gibbs(
+            GibbsConditional(:m, c -> Normal(m_true, 0)),
+            GibbsConditional(:λ, cond_λ))
+        chain = sample(inverse_gdemo(x), alg, 10_000)
+        check_numerical(chain, [:m, :λ], [m_true, λ_true], atol=0.2)
+        
+        alg = Gibbs(
+            GibbsConditional(:m, cond_m),
+            GibbsConditional(:λ, cond_λ))
+        chain = sample(inverse_gdemo(x), alg, 10_000)
+        check_numerical(chain, [:m, :λ], [m_true, λ_true], atol=0.2)
+    end
+
+    let π = [0.5, 0.5],
+        K = length(π),
+        m = 0.5,
+        λ = 2.0,
+        σ = 0.1,
+        x = [σ .* randn(10); 1 .+ σ .* randn(10)],
+        N = length(x)
+        
+        @model mixture(x) = begin
+            μ ~ arraydist(Normal.(fill(m, K), fill(λ, K)))
+            z ~ arraydist(Categorical.(fill(π, N)))
+            x ~ arraydist(Normal.(μ[z], σ))
+            return x
+        end
+
+        # see http://www.cs.columbia.edu/~blei/fogm/2015F/notes/mixtures-and-gibbs.pdf
+        function cond_z(c)
+            function mixtureweight(x)
+                p = π .* pdf.(Normal.(c.μ, σ), Ref(x))
+                return p ./ sum(p)
+            end
+            return arraydist(Categorical.(mixtureweight.(x)))
+        end
+
+        function cond_μ(c)
+            z = c.z
+            n = [count(z .== k) for k = 1:K]
+
+            # If there were no observations assigned to center `k`, `n[k] == 0`, and
+            # we use the prior instead.
+            x_bar = [(n[k] != 0) ? (sum(x[z .== k]) / n[k]) : m for k = 1:K]
+            λ_hat = [(n[k] != 0) ? inv(n[k] / σ^2 + 1/λ^2) : λ for k = 1:K]
+            μ_hat = [(n[k] != 0) ? x_bar[k] * (n[k] / σ^2) * λ_hat[k] : m for k = 1:K]
+
+            return arraydist(Normal.(μ_hat, λ_hat))
+        end
+
+        Random.seed!(100)
+        alg = Gibbs(GibbsConditional(:z, cond_z), GibbsConditional(:μ, cond_μ))
+        chain = sample(mixture(x), alg, 10000)
+        
+        check_numerical(chain, [:μ], [[0, 1]], atol=0.1)
+
+        # we can't be sure about the order of cluster assignment, so we check both
+        ẑ = dropdims(mean(chain[:z].value, dims=1), dims=(1, 3))
+        z_true = [fill(1, 10); fill(2, 10)]
+        @test isapprox(ẑ, z_true, atol=0.2, rtol=0.0) || isapprox(ẑ, z_true[[11:20; 1:10]], atol=0.2, rtol=0.0)
```

Comment:
```suggestion
        @test isapprox(ẑ, z_true, atol=0.2, rtol=0.0) || isapprox(ẑ, z_true[[11:20; 1:10]], atol=0.2, rtol=0.0)
        
        Random.seed!(100)
        alg = Gibbs(GibbsConditional(:z, cond_z), HMC(0.05, 4, :μ,))
        chain = sample(mixture(x), alg, 10000)

        check_numerical(chain, [:μ], [[0, 1]], atol=0.1)

        # we can't be sure about the order of cluster assignment, so we check both
        ẑ = dropdims(mean(chain[:z].value, dims=1), dims=(1, 3))
        z_true = [fill(1, 10); fill(2, 10)]
        @test isapprox(ẑ, z_true, atol=0.2, rtol=0.0) || isapprox(ẑ, z_true[[11:20; 1:10]], atol=0.2, rtol=0.0)
```

Reply:
On my system, this works for Tracker, but not ReverseDiff or ForwardDiff -- so I suspect AD is broken?  What would be an appropriate non-Hamiltonian sampler to compare HMC to in this case?

Reply:
@mohamed82008 do you know why this would be broken for ForwardDiff?

Reply:
You could use `MH`, but ideally we want to test against `HMC` as this will be the most frequently used scenario.

Reply:
MH passes.  Which strongly indicates that AD is in fact the problem, I'd say.  I'll have a look at the implementation again to see whether I find anything suspicious.

Reply:
Ah, I'm using `foldl(vcat, arrays)`, which is discussed [over in ReverseDiff](https://github.com/JuliaDiff/ReverseDiff.jl/issues/122#issuecomment-609075734) already.

Reply:
Or maybe not... `f(xs) = sum(foldl(vcat, [xs[1:2]; xs[3:3]]))` does have a correct ReverseDiff gradient.

Reply:
I missread the test results, it's actually Reversediff and Tracker that fail, and ForwardDiff which works.  Both get the value of the first entry of \mu right, but the second one is off to the negative (wiggling around -3).